File: reddit_post.py
import requests
import random
import sys
from dataclasses import dataclass
from typing import Optional

REDDIT_API = "https://www.reddit.com/r/"
USER_AGENT = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/142.0.0.0 Safari/537.36"

@dataclass
class RedditPost:
    subreddit: str
    title: str
    description: str
    url: str
    media_url: str
    author: str
    score: int
    upvotes: int
    downvotes: int

def get_random_reddit_post(sub_reddit: str) -> Optional[RedditPost]:
    url = f"{REDDIT_API}{sub_reddit}.json"
    headers = {
        "User-Agent": USER_AGENT,
        "Accept": "application/json"
    }

    try:
        response = requests.get(url, headers=headers)
        response.raise_for_status()
        
        data = response.json()
        posts_data = data["data"]["children"]
        total_posts = len(posts_data)
        
        print(f"Total posts found: {total_posts}")

        if total_posts == 0:
            print("No posts found in the subreddit.")

        random_index = random.randint(0, total_posts - 1)
        print(f"Selected post index: {random_index}")

        post = posts_data[random_index]["data"]

        return RedditPost(
            title=post.get("title"),
            subreddit=post.get("subreddit"),
            description=post.get("selftext"),
            media_url=post.get("url"),
            url=f"https://reddit.com{post.get('permalink')}",
            author=post.get("author"),
            score=post.get("score"),
            upvotes=post.get("ups"),
            downvotes=post.get("downs")
        )
    except Exception as error:
        print(f"Error fetching data: {error}")

if __name__ == "__main__":
    sub_reddit = input("Sub reddit to get a random post: ").strip()
    
    random_post = get_random_reddit_post(sub_reddit)

    if random_post is None:
        print("Failed to retrieve a post.")
    else:
        print(f"Title: {random_post.title}")
        print(f"Subreddit: {random_post.subreddit}")
        print(f"Author: {random_post.author}")
        print(f"Score: {random_post.score}")
        print(f"Upvotes: {random_post.upvotes}")
        print(f"Downvotes: {random_post.downvotes}")
        print(f"Media URL: {random_post.media_url}")
        print(f"URL: {random_post.url}")

        if random_post.description:
            print(f"Description: {random_post.description}")
        else:
            print("No description available.")

File: reddit_post.nim
import std/strutils
import httpclient
import json
import random

const redditAPI = "http://www.reddit.com/r/"

stdout.write("Sub reddit to get a random post: ")
stdout.flushFile()
let subReddit: string = readLine(stdin)

randomize() # this make a new seed for random number generation

type 
    RedditPost = object
        subreddit: string
        title: string
        description: string
        url: string
        mediaUrl: string
        author: string
        score: int
        upvotes: int
        downvotes: int

proc getRandomRedditPost(subReddit: string): RedditPost =
    let url = redditAPI & subReddit & ".json"
    var client = newHttpClient(
      headers = newHttpHeaders({
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/142.0.0.0 Safari/537.36",
        "Accept": "application/json"
      })
    )

    let response = client.getContent(url)
    
    var posts = parseJson(response)
    let postsData = posts["data"]["children"]
    let totalPosts = postsData.len()
    echo "Total posts found: ", totalPosts

    if totalPosts == 0:
        echo "No posts found in the subreddit."
        quit(1)

    let randomIndex = rand(totalPosts)
    echo "Selected post index: ", randomIndex

    let post = postsData[randomIndex]["data"]
    result.title = post["title"].getStr()
    result.subreddit = post["subreddit"].getStr()
    result.description = post["selftext"].getStr()
    result.mediaUrl = post["url"].getStr()
    result.url = "https://reddit.com" & post["permalink"].getStr()
    result.author = post["author"].getStr()
    result.score = post["score"].getInt()
    result.upvotes = post["ups"].getInt()
    result.downvotes = post["downs"].getInt()

let randomPost = getRandomRedditPost(subReddit)

echo "Title: ", randomPost.title # using the comma avoids creating a third string for concatenation...  i think
echo "Subreddit: ", randomPost.subreddit
echo "Author: ", randomPost.author
echo "Score: ", randomPost.score
echo "Upvotes: ", randomPost.upvotes
echo "Downvotes: ", randomPost.downvotes
echo "Media URL: ", randomPost.mediaUrl
echo "URL: ", randomPost.url

if randomPost.description.len > 0:
    echo "Description: ", randomPost.description
else:
    echo "No description available."